### PR TITLE
Show recommended policy for new deployments

### PR DESCRIPTION
**Pull request by vishnusomank**
- added deployment watcher to generate recommended policies when a new deployment is added

Signed-off-by: Vishnu Soman <vishnu@accuknox.com>

### CODE REVIEW

File: src/recommendpolicy/helperFunctions.go [hunk 1]
```diff
@@ -118,3 +123,45 @@ func addPolicyRule(policy *types.KnoxSystemPolicy, r *types.KnoxSystemSpec) {
 	}
 
 }
+
+func initDeploymentWatcher() {
+	clientset := cluster.ConnectK8sClient()
+	watcher, err := clientset.AppsV1().Deployments("").Watch(context.TODO(), metav1.ListOptions{})
+
+	if err != nil {
+		log.Error().Msg(err.Error())
```
Comment (nyrahul) [inline src/recommendpolicy/helperFunctions.go:133]: return

File: src/recommendpolicy/recommendPolicy.go [hunk 2]
```diff
@@ -131,16 +137,45 @@ func RecommendPolicyMain() {
...
 			}
 		}
 	}
+}
+
+func generateHardenPolicy(name, namespace string, labels LabelMap) {
+	log.Info().Msgf("Generating hardening policy for deployment: %v in namespace: %v", name, namespace)
+	policies, err := generatePolicy(name, namespace, labels)
+	if err != nil {
+		log.Error().Msg(err.Error())
```
Comment (nyrahul) [inline src/recommendpolicy/recommendPolicy.go:158]: return

**Review by nyrahul — CHANGES_REQUESTED** [review]

File: src/libs/consumer.go [hunk 3]
```diff
@@ -92,6 +92,7 @@ func (pc *PolicyStore) Publish(policy *types.PolicyYaml) {
 
 	for consumer := range pc.Consumers {
 		if matchPolicyYaml(policy, consumer) {
+			log.Info().Msgf("Publishing policy %v", policy.Name)
```
Comment (seswarrajan) [inline src/libs/consumer.go:95]: Is this message necessary ?
The purpose of this message is for debug ?

**Review by nyrahul — APPROVED** [review]

File: src/recommendpolicy/helperFunctions.go [hunk 4]
```diff
@@ -118,3 +124,46 @@ func addPolicyRule(policy *types.KnoxSystemPolicy, r *types.KnoxSystemSpec) {
...
+		case watch.Added:
+			for _, data := range DeployNsName {
+				if data.Name == event.Object.(*v1.Deployment).Name && data.Namespace == event.Object.(*v1.Deployment).Namespace {
+					found = true
+					break
+				}
+			}
+			if !found {
+				log.Info().Msgf("Found Deployment %v in %v namespace", event.Object.(*v1.Deployment).Name, event.Object.(*v1.Deployment).Namespace)
+				generateHardenPolicy(event.Object.(*v1.Deployment).Name, event.Object.(*v1.Deployment).Namespace, event.Object.(*v1.Deployment).Spec.Template.Labels)
```
Comment (seswarrajan) [inline src/recommendpolicy/helperFunctions.go:151]: Are we extracting recommended policies based only on deployment name & labels
What if the same contains multiple containers ?? We are not getting recommended policies for containers ?

**Review by seswarrajan — COMMENTED** [review]
Please check how much CPU/mem is utilised since we are using a watcher .

**Pull request merged**